fxa-client/sdks/android/library/src/main/java/org/mozilla/fxa_client/internal/FirefoxAccount.kt

```diff
@@ -0,0 +1,149 @@
+        }.getAndConsumeString()
+    }
+
+    fun beginPairingFlow(pairingUrl: String, scopes: Array<String>): String {
+        val scope = scopes.joinToString(" ")
+        return rustCall { e ->
+            FxaClient.INSTANCE.fxa_begin_pairing_flow(validPointer(), pairingUrl, scope, e)
+        }.getAndConsumeString()
+    }
+
+    /**
+     * Fetches the profile object for the current client either from the existing cached account,
+     * or from the server (requires the client to have access to the profile scope).
+     *
+     * @param ignoreCache Fetch the profile information directly from the server
+     * @return [Profile] representing the user's basic profile info
+     */
+    fun getProfile(ignoreCache: Boolean): Profile {
+        return Profile(rustCall { e ->
+            FxaClient.INSTANCE.fxa_profile(validPointer(), ignoreCache, e)
+        })
+    }
+
+    /**
+     * Convenience method to fetch the profile from a cached account by default, but fall back
+     * to retrieval from the server.
+     *
+     * @return [Profile] representing the user's basic profile info
+     */
+    fun getProfile(): Profile {
+        return getProfile(false)
+    }
+
+    /**
+     * Fetches the token server endpoint, for authentication using the SAML bearer flow.
+     */
+    fun getTokenServerEndpointURL(): String? {
+        return rustCall { e ->
+            FxaClient.INSTANCE.fxa_get_token_server_endpoint_url(validPointer(), e)
+        }.getAndConsumeString()
+    }
+
+    /**
+     * Fetches keys for encryption/decryption of Firefox Sync data.
+     */
+    fun getSyncKeys(): SyncKeys {
+        return SyncKeys(rustCall { e ->
+            FxaClient.INSTANCE.fxa_get_sync_keys(validPointer(), e)
+        })
+    }
+
+    /**
+     * Authenticates the current account using the code and state parameters fetched from the
+     * redirect URL reached after completing the sign in flow triggered by [beginOAuthFlow].
+     *
+     * Modifies the FirefoxAccount state.
+     */
+    fun completeOAuthFlow(code: String, state: String): OAuthInfo {
+        return OAuthInfo(rustCall { e ->
+            FxaClient.INSTANCE.fxa_complete_oauth_flow(validPointer(), code, state, e)
+        })
+    }
+
+    /**
+     * Fetches a new access token for the desired scopes using an internally stored refresh token.
+     *
+     * @param scopes List of OAuth scopes for which the client wants access
+     * @return [OAuthInfo] that stores the token, along with its scopes and keys when complete
+     * @throws FxaException.Unauthorized if the token could not be retrieved (eg. expired refresh token)
+     */
+    fun getOAuthToken(scopes: Array<String>): OAuthInfo {
+        val scope = scopes.joinToString(" ")
+        return OAuthInfo(nullableRustCall { e ->
+            FxaClient.INSTANCE.fxa_get_oauth_token(validPointer(), scope, e)
+        } ?: throw FxaException.Unauthorized("Restart OAuth flow"))
```

Comment:
@eoger The FxA code returns `Result<Option<OAuthInfo>>`, but i'm not 100% sure when it returns `Ok(None)`. ISTM like it should emit an error in rust instead?

Reply:
This is an expected result as this function is more like `getCachedOAuthToken`.

Reply:
Should it return null then instead of throwing? Should it's name be changed (at least on the kotlin side)?

Reply:
I'm thinking it should be null yes. I'm ok with the current name but if you think it's valuable 👍